client/network/sync/src/lib.rs

```diff
@@ -2175,9 +2179,20 @@ where
 	}
 
 	fn poll_pending_responses(&mut self, cx: &mut std::task::Context) -> Poll<ImportResult<B>> {
-		while let Poll::Ready(Some((id, request, response))) =
-			self.pending_responses.poll_next_unpin(cx)
-		{
+		let ready_responses = self
+			.pending_responses
```

Comment:
Just a thought, this hashmap will be iterated quite often. Maybe it's worth looking at [indexmap](https://docs.rs/indexmap/latest/indexmap/) for this for iteration performance? But I am not sure how many elements there are typically in `pending_responses`, so maybe this is a premature optimization.

Reply:
Thanks for the suggestion. When syncing was still in `Protocol`, responses were stored the same as they're done in this PR, i.e., iterating over all peers. There should a maximum of 40 responses (number of peers) in the hashmap when major syncing and only a few when keep-up syncing. But I agree, this could definitely be better and I will look into it in the near future: https://github.com/paritytech/substrate/issues/13962